others/serialization.cpp

```diff
@@ -0,0 +1,190 @@
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <cassert> 
+
+/**
+ * @file
+ * @brief A simple Serializer and Deserializer utility for fundamental data
+ * types and strings.
+ */
+
+class Serializer {
+ public:
+    /**
+     * @brief Serializes fundamental data types (like int, float, double, etc.)
+     * to a binary file.
+     * @tparam T The type of the data to be serialized.
+     * @param out The output stream (std::ofstream).
+     * @param data The data to be serialized.
+     *
+     * @note This function only works for fundamental types (primitives).
+     */
+    template <typename T>
+    static void serialize(std::ofstream &out, const T &data) {
+        static_assert(std::is_fundamental<T>::value, "Non-fundamental types are not allowed for this function!");
+        out.write(reinterpret_cast<const char *>(&data), sizeof(T));
+    }
+
+    /**
+     * Serializes a string to a binary file.
+     * @param out The output stream (std::ofstream).
+     * @param data The string to be serialized.
+     *
+     * The string is serialized by first storing its length, followed by the
+     * content.
+     */
+    static void serialize(std::ofstream &out, const std::string &data) {
+        ssize_t length = data.size();
+        serialize(out, length);           // Serialize the length of the string.
+        out.write(data.c_str(), length);  // Serialize the string characters.
+        out.put('|');  // Add a delimiter to denote the end of the string.
```

Comment:
could the delimiter be made a `class member` defined during the class construction?

Reply:
If you mean something like this for both the classes
```cpp
class Serializer
public:
     char delimiter = '|' //since c++ 11
     //rest of the code
};
```

We cannot do that as we wont be able to access the delimiter inside the static class methods. For this to work we have to make the ``char delimiter`` as static variable and in future the user can define custom delimiters, but this can cause confusion as if the user accidentally uses different delimiters in a single file for serialization for deserialization.